crypto/keyring.go

```diff
@@ -347,12 +173,57 @@ func FilterExpiredKeys(contactKeys []*KeyRing) (filteredKeys []*KeyRing, err err
 }
 
 // FirstKey returns a KeyRing with only the first key of the original one
-func (keyRing *KeyRing) FirstKey() *KeyRing {
+func (keyRing *KeyRing) FirstKey() (*KeyRing, error) {
 	if len(keyRing.entities) == 0 {
-		return nil
+		return nil, errors.New("gopenpgp: No key available in this keyring")
 	}
 	newKeyRing := &KeyRing{}
 	newKeyRing.entities = keyRing.entities[:1]
 
-	return newKeyRing
+	return newKeyRing.Copy()
+}
+
+// Copy creates a deep copy of the keyring
+func (keyRing *KeyRing) Copy() (*KeyRing, error) {
+	newKeyRing := &KeyRing{}
+
+	entities := make([]*openpgp.Entity, len(keyRing.entities))
+	for id, entity := range keyRing.entities {
+		var buffer bytes.Buffer
+		var err error
+
+		if entity.PrivateKey == nil {
+			err = entity.Serialize(&buffer)
+		} else {
+			err = entity.SerializePrivateNoSign(&buffer, nil)
```

Comment:
Perhaps this code should be moved to the method on Key, and we should be calling that?
(I know, I know, it's not an API/interface comment.. feel free to ignore until later)

Reply:
Would you just clone it by cloning all the keys in the keyring, instead of having this "duplicate" code?

Reply:
In any case it is postponed to 2.0.1 :)